### PR TITLE
Added JIT_Stelem_Ref helper function for Linux

### PR DESCRIPTION
**Pull request by helloguo**
This asm helper function can replace the portable version and achieve better performance on Ubuntu 14.04 and Mac OS X 10.11.

### CODE REVIEW

**Comment by jkotas** [comment]
@janvorli PTLA

File: src/vm/amd64/jithelpers_fast.S [hunk 1]
```diff
@@ -329,3 +329,152 @@ LEAF_ENTRY JIT_ByRefWriteBarrier, _TEXT
...
+
+// TODO: put definition for this in asmconstants.h
+#define CanCast 1
+
+//__declspec(naked) void F_CALL_CONV JIT_Stelem_Ref(PtrArray* array, unsigned idx, Object* val)
+.balign 16
+LEAF_ENTRY JIT_Stelem_Ref, _TEXT
+        // check for null PtrArray*
+        test    rdi, rdi
+        je      5f // ThrowNullReferenceException
```
Comment (janvorli) [inline src/vm/amd64/jithelpers_fast.S]: Could you please use LOCAL_LABEL(ThrowNullReferenceException) instead the numeric label and a comment? The same at all other place.

File: src/vm/amd64/jithelpers_fast.S [hunk 2]
```diff
@@ -329,3 +329,152 @@ LEAF_ENTRY JIT_ByRefWriteBarrier, _TEXT
         add     rsi, 8h
         ret
 LEAF_END JIT_ByRefWriteBarrier, _TEXT
+
+#include "asmconstants.h"
```
Comment (janvorli) [inline src/vm/amd64/jithelpers_fast.S]: Please move this to the beginning of the file right after the other include.

File: src/vm/amd64/jithelpers_fast.S [hunk 3]
```diff
@@ -329,3 +329,152 @@ LEAF_ENTRY JIT_ByRefWriteBarrier, _TEXT
...
+        mov     rdi, rdx
+
+        // TypeHandle::CastResult ObjIsInstanceOfNoGC(Object *pElement, TypeHandle toTypeHnd)
+        call    C_FUNC(ObjIsInstanceOfNoGC)
+
+        mov     rdi, [rbp - 0x08]
+        mov     rsi, [rbp - 0x10]
+        mov     rdx, [rbp - 0x18]
+        
+        mov     rsp, rbp
```
Comment (janvorli) [inline src/vm/amd64/jithelpers_fast.S]: It would be nice to replace this whole block (4 lines) with macro, named e.g. "RESET_FRAME" or something  similar and use it also at the other place before.

File: src/vm/amd64/asmconstants.h [hunk 4]
```diff
@@ -663,7 +663,9 @@ ASMCONSTANTS_C_ASSERT(OFFSETOF__TransitionBlock__m_argumentRegisters == offsetof
 
 #undef ASMCONSTANTS_RUNTIME_ASSERT
 #undef ASMCONSTANTS_C_ASSERT
+#ifndef UNIX_AMD64_ABI
 #undef DBG_FRE
```
Comment (janvorli) [inline src/vm/amd64/asmconstants.h]: What is the purpose of this change?

Comment (helloguo) [inline src/vm/amd64/asmconstants.h]: DBG_FRE is used by OFFSETOF__MethodTable__m_ElementType, which is used by JIT_Stelem_Ref. If this micro is undef here, the build process complains "unknown token in expression".

File: src/vm/amd64/jithelpers_fast.S (hunk 1 above)
Comment (helloguo) [inline src/vm/amd64/jithelpers_fast.S]: Yes, it's changed to LOCAL_LABEL(lable_name).

File: src/vm/amd64/jithelpers_fast.S (hunk 2 above)
Comment (helloguo) [inline src/vm/amd64/jithelpers_fast.S]: Done.

File: src/vm/amd64/jithelpers_fast.S (hunk 3 above)
Comment (helloguo) [inline src/vm/amd64/jithelpers_fast.S]: Yes, does RESET_FRAME_WITH_RBP work?

File: src/vm/amd64/asmconstants.h (hunk 4 above)
Comment (janvorli) [inline src/vm/amd64/asmconstants.h]: That is strange, the assembler should get a fully preprocessed file, so it should not encounter the DBG_FRE. Could you please show me the full error message?

Comment (helloguo) [inline src/vm/amd64/asmconstants.h]: /tmp/jithelpers_fast-77dca9.s:1040:32: error: unknown token in expression
        mov rcx, [r10 + DBG_FRE(0x38, 0x30)]
                               ^
make[2]: *** [src/vm/wks/CMakeFiles/cee_wks.dir/__/amd64/jithelpers_fast.S.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [src/vm/wks/CMakeFiles/cee_wks.dir/all] Error 2
make: *** [all] Error 2
Failed to build coreclr components.

I build it on Ubuntu 14.04.

Comment (janvorli) [inline src/vm/amd64/asmconstants.h]: Ok, sounds like a quirk of the preprocessor then.

**Comment by janvorli** [comment]
LGTM

**Comment by helloguo** [comment]
From the test report, Ubuntu x64 Checked Build and Test is failed due to "There is not enough space on the disk". Is there a way to re-run this test?

**Comment by janvorli** [comment]
@dotnet-bot test Ubuntu x64 Checked Build and Test please

**Comment by helloguo** [comment]
@dotnet-bot can you run Ubuntu x64 Checked Build and Test? From the report, the last run has error: "Jenkins Remoting Error" and "Build went offline during the build".

**Comment by helloguo** [comment]
@dotnet-bot test Ubuntu x64 Checked Build and Test please

**Comment by jkotas** [comment]
Thanks!
